extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQFaultsTest.java

```diff
@@ -356,4 +359,44 @@ public void testTooManyInputFiles() throws IOException
         .setExpectedMSQFault(new TooManyInputFilesFault(numFiles, Limits.MAX_INPUT_FILES_PER_WORKER, 2))
         .verifyResults();
   }
+
+  @Test
+  public void testUnionAllIsDisallowed()
+  {
+    final RowSignature rowSignature =
+        RowSignature.builder().add("__time", ColumnType.LONG).build();
+    testIngestQuery()
+        .setSql("SELECT * FROM foo\n"
+                + "UNION ALL\n"
+                + "SELECT * FROM foo\n")
+        .setExpectedRowSignature(rowSignature)
+        .setExpectedDataSource("foo1")
+        .setExpectedMSQFault(QueryNotSupportedFault.instance())
+        .verifyResults();
+  }
+
+  @Test
+  public void testUnionAllIsDisallowedWhilePlanning()
+  {
+    // This results in a planning error however the planning error isn't an accurate representation of the actual error
+    // because Calcite rewrites it using CoreRules.UNION_TO_DISTINCT, which plans it using Union Datasource.
+    // However, this fails since the column names mismatch. Once MSQ is able to support Union datasources, the planning
+    // error would become an accurate representation of the error.
+    testIngestQuery()
+        .setSql(
+            "INSERT INTO druid.dst "
+            + "SELECT dim2, dim1, m1 FROM foo2 "
+            + "UNION ALL "
+            + "SELECT dim1, dim2, m1 FROM foo "
+            + "PARTITIONED BY ALL TIME")
+        .setExpectedValidationErrorMatcher(
+            new DruidExceptionMatcher(
+                DruidException.Persona.ADMIN,
+                DruidException.Category.INVALID_INPUT,
+                "general"
+            ).expectMessageIs("Query planning failed for unknown reason, our best guess is this "
```

Comment:
Where is this error coming from? Looking at the code for the union rule, I would think it can't happen, because it's generated by `isCompatible`, which isn't called when the `ALLOW_TOP_LEVEL_UNION_ALL` feature is missing. The error should be something about `UNION ALL` being unsupported for this engine.

Reply:
It's because top-level union all is disallowed, so it tries to plan the query using the [UnionDataSourceRule](https://github.com/apache/druid/blob/master/sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidUnionDataSourceRule.java), goes into the `isCompatible` then, and then rewrites the already set planning error.

Reply:
I'll clarify the two test cases that I added

Reply:
Added comments and renamed the test cases. Hope they clarify the confusion